include/nbl/builtin/hlsl/bxdf/reflection/lambertian.hlsl

```diff
@@ -0,0 +1,176 @@
+    using quotient_pdf_type = sampling::quotient_and_pdf<spectral_type, scalar_type>;
+
+    using params_isotropic_t = LambertianParams<LS, Iso, scalar_type>;
+    using params_anisotropic_t = LambertianParams<LS, Aniso, scalar_type>;
+
+
+    static this_t create()
+    {
+        this_t retval;
+        // nothing here, just keeping in convention with others
+        return retval;
+    }
+
+    static this_t create(NBL_CONST_REF_ARG(SBxDFCreationParams<scalar_type, spectral_type>) params)
+    {
+        return create();
+    }
+
+    void init(NBL_CONST_REF_ARG(SBxDFCreationParams<scalar_type, spectral_type>) params)
+    {
+        // do nothing
+    }
+
+    scalar_type __eval_pi_factored_out(scalar_type maxNdotL)
+    {
+        return maxNdotL;
+    }
+
+    scalar_type eval(NBL_CONST_REF_ARG(params_isotropic_t) params)
+    {
+        return __eval_pi_factored_out(params.getNdotL()) * numbers::inv_pi<scalar_type>;
+    }
+    scalar_type eval(NBL_CONST_REF_ARG(params_anisotropic_t) params)
+    {
+        return __eval_pi_factored_out(params.getNdotL()) * numbers::inv_pi<scalar_type>;
+    }
+
+    sample_type generate_wo_clamps(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(vector<scalar_type, 2>) u)
+    {
+        ray_dir_info_type L;
+        L.direction = sampling::ProjectedHemisphere<scalar_type>::generate(u);
+        return sample_type::createFromTangentSpace(interaction.getTangentSpaceV(), L, interaction.getFromTangentSpace());
+    }
+
+    sample_type generate(NBL_CONST_REF_ARG(anisotropic_interaction_type) interaction, NBL_CONST_REF_ARG(vector<scalar_type, 2>) u)
+    {
+        return generate_wo_clamps(interaction, u);
+    }
+
+    sample_type generate(NBL_CONST_REF_ARG(isotropic_interaction_type) interaction, NBL_CONST_REF_ARG(vector<scalar_type, 2>) u)
+    {
+        return generate_wo_clamps(anisotropic_interaction_type::create(interaction), u);
+    }
+
+    scalar_type pdf(NBL_CONST_REF_ARG(params_isotropic_t) params)
+    {
+        return sampling::ProjectedHemisphere<scalar_type>::pdf(params.getNdotL());
+    }
+    scalar_type pdf(NBL_CONST_REF_ARG(params_anisotropic_t) params)
+    {
+        return sampling::ProjectedHemisphere<scalar_type>::pdf(params.getNdotL());
+    }
+
+    quotient_pdf_type quotient_and_pdf(NBL_CONST_REF_ARG(params_isotropic_t) params)
+    {
+        sampling::quotient_and_pdf<scalar_type, scalar_type> qp = sampling::ProjectedHemisphere<scalar_type>::template quotient_and_pdf(params.getNdotL());
+        return quotient_pdf_type::create(hlsl::promote<spectral_type>(qp.quotient), qp.pdf);
+    }
+    quotient_pdf_type quotient_and_pdf(NBL_CONST_REF_ARG(params_anisotropic_t) params)
+    {
+        sampling::quotient_and_pdf<scalar_type, scalar_type> qp = sampling::ProjectedHemisphere<scalar_type>::template quotient_and_pdf(params.getNdotL());
+        return quotient_pdf_type::create(hlsl::promote<spectral_type>(qp.quotient), qp.pdf);
```

Comment:
imho `quotient_pdf_type` should have a `static create` from a scalar which already calls the `promote` for you (or even a `create` from a related quotient_and_pdf type where the quotient type is `vector<T,1>`)

will save you some typing across all BxDFs